### PR TITLE
WIP: nuclear gradients

### PR DESCRIPTION
**Pull request by hatemhelal**

### CODE REVIEW

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    primitive_op: Callable, atom_index: int, a: Primitive, b: Primitive
+) -> Float3:
+    """Generic gradient of a one-electron integral with respect the atom_index center"""
+    axes = jnp.arange(3)
+    lhs_p1 = vmap(a.offset_lmn, (0, None))(axes, 1)
+    t1 = 2 * a.alpha * vmap(primitive_op, (0, None))(lhs_p1, b)
+
+    lhs_m1 = vmap(a.offset_lmn, (0, None))(axes, -1)
+    t2 = jnp.where(a.lmn > 0, a.lmn, jnp.zeros_like(a.lmn))
+    t2 *= vmap(primitive_op, (0, None))(lhs_m1, b)
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:24]: I realise this is sacrilege, but I wonder if a python loop over 0,1,2 might be easier to read, and ultimately be no slower, as it will be vectorisable by a higher level vmap.
E.g. something like (artist's impression)
```
grad = jnp.zero(3)
for axis in range(3):
  t1 = 2 * a.alpha * primitive_op(a.offset_lmn[axis], 1, b)
  t2 = a.offset_lmn[axis] * primitive_op(a.offset_lmn[axis], -1, b)
  grad = grad.at(axis).set(t1 - t2)

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+from .basis import Basis
+from .integrals import _overlap_primitives, _kinetic_primitives
+from .orbital import batch_orbitals
+from .primitive import Primitive
+from .types import Float3, Float3xNxN
+
+
+def grad_primitive_integral(
+    primitive_op: Callable, atom_index: int, a: Primitive, b: Primitive
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:15]: What's the signature of this Callable?  It's great to point to an example, even from this more generic code to less generic.

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    """Generic gradient of a one-electron integral with respect the atom_index center"""
+    axes = jnp.arange(3)
+    lhs_p1 = vmap(a.offset_lmn, (0, None))(axes, 1)
+    t1 = 2 * a.alpha * vmap(primitive_op, (0, None))(lhs_p1, b)
+
+    lhs_m1 = vmap(a.offset_lmn, (0, None))(axes, -1)
+    t2 = jnp.where(a.lmn > 0, a.lmn, jnp.zeros_like(a.lmn))
+    t2 *= vmap(primitive_op, (0, None))(lhs_m1, b)
+    grad_out = t1 - t2
+    return jnp.where(a.atom_index == atom_index, grad_out, jnp.zeros_like(grad_out))
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:26]: This is a lot of computation to then get multiplied by zero -- can we not scatter/gather this from a higher level?   I.e. ensure this code is only called where atom_index = a.atom_index?

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def grad_integrate(b: Basis, primitive_op: Callable) -> Float3xNxN:
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:29]: Maybe `b` -> `basis`, as `b` is used in a lot of places to mean a Primitive?

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    """Generic gradient of one-electron integrals over the basis set"""
+
+    def take_primitives(indices):
+        p = tree_map(lambda x: jnp.take(x, indices, axis=0), primitives)
+        c = jnp.take(coefficients, indices)
+        return p, c
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:35]: I like to define these local lambdas after the variables to which they refer have been defined.  It puts the computation closer to the point of use.

(i.e. move it after line 37)

OTOH, `batch_orbitals` is already doing a lot of Python list comprehensions, so it might be clearer and the same complexity to treat `b` as a list of lists of primitives, and just inline the list comprehensions here?

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    primitives, coefficients, orbital_index = batch_orbitals(b.orbitals)
+    ii, jj = jnp.meshgrid(*[jnp.arange(b.num_primitives)] * 2, indexing="ij")
+    lhs, cl = take_primitives(ii.reshape(-1))
+    rhs, cr = take_primitives(jj.reshape(-1))
+
+    op = vmap(primitive_op, (None, 0, 0))
+    op = jit(vmap(op, (0, None, None)))
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:43]: Does this jit change much?  It feels a bit out of place, but I get that it might be needed.

File: pyscf_ipu/experimental/nuclear_gradients.py (hunk 1 above)
Comment (awf) [reply pyscf_ipu/experimental/nuclear_gradients.py:24]: Or if that turns out to be slower, it might serve as reference code in another function so testing can compare them.

File: pyscf_ipu/experimental/nuclear_gradients.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    atom_indices = jnp.arange(b.structure.num_atoms)
+    out = op(atom_indices, lhs, rhs)
+    out = jnp.sum(out, axis=0)
+
+    out = cl * cr * out.T
+    out = out.reshape(3, b.num_primitives, b.num_primitives)
+    out = segment_sum(jnp.rollaxis(out, 1), orbital_index)
+    out = segment_sum(jnp.rollaxis(out, -1), orbital_index)
+
+    return jnp.rollaxis(out, -1)
```
Comment (awf) [inline pyscf_ipu/experimental/nuclear_gradients.py:92]: Is this now supplanted by the generic one above?

File: pyscf_ipu/experimental/primitive.py [hunk 8]
```diff
@@ -14,6 +15,7 @@ class Primitive:
     alpha: float = 1.0
     lmn: Int3 = jnp.zeros(3, dtype=jnp.int32)
     norm: Optional[float] = None
+    atom_index: Optional[int] = None
```
Comment (awf) [inline pyscf_ipu/experimental/primitive.py:18]: This is a back-pointer into the list of which this is a member?
That's also a hint that the `where(a.atom_index == atom_index)` could be lifted.

File: pyscf_ipu/experimental/nuclear_gradients.py (hunk 7 above)
Comment (hatemhelal) [reply pyscf_ipu/experimental/nuclear_gradients.py]: yes sorry forgot to push that commit...

File: pyscf_ipu/experimental/nuclear_gradients.py (hunk 1 above)
Comment (hatemhelal) [reply pyscf_ipu/experimental/nuclear_gradients.py]: trying out a list comprehension like this:

```python
    t1 = [primitive_op(a.offset_lmn(ax, 1), b) for ax in range(3)]
    t2 = [primitive_op(a.offset_lmn(ax, -1), b) for ax in range(3)]
    grad_out = 2 * a.alpha * jnp.stack(t1) - a.lmn * jnp.stack(t2)
    return grad_out
```
better or worse?

Comment (awf) [reply pyscf_ipu/experimental/nuclear_gradients.py]: Definitely easier to read - perf impact ?
